ref-farming/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+/*!
+* Ref-Farming
+*
+* lib.rs is the main entry point.
+*/
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::{ValidAccountId};
+use near_sdk::collections::{LookupMap, UnorderedMap};
+use near_sdk::{env, near_bindgen, Balance, AccountId, PanicOnDefault};
+
+use crate::farm::{Farm, FarmId};
+use crate::farm_seed::{FarmSeed, SeedId};
+use crate::farmer::Farmer;
+
+// for simulator test
+pub use crate::simple_farm::HRSimpleFarmTerms;
+pub use crate::view::FarmInfo;
+
+
+mod utils;
+mod errors;
+mod farmer;
+mod token_receiver;
+mod farm_seed;
+mod farm;
+mod simple_farm;
+mod storage_impl;
+
+mod actions_of_farm;
+mod actions_of_seed;
+mod actions_of_reward;
+mod view;
+
+/// sodu module is used to debug and testing,
+/// remove this module in release version
+mod sudo;
+
+near_sdk::setup_alloc!();
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+
+    // owner of this contract
+    owner_id: AccountId,
```

Comment:
does this contract need an owner? it doesn't seem like it's used anywhere

Reply:
The owner is used for the next move. Cause you can see the farm's life cycle is incomplete, we can only create a farm and it will last for ever, like the simple-pool in ref-finance. So, the next version of this farming contract, from my view, is to address that. When all the reward of a farm is distributed and claimed out, then it could be removed from contract states. The way to do that, is to have kind of owner of this contract, act as a backend operator, have the rights to clean outdated farms. And like ref-finance, I am thinking adding upgrade function to the contract, which also need the exist of owner, I think.